### ISSUE TITLE
Clicking away from menu should close it

### ISSUE THREAD

**Issue by SabreCat** [HabitRPG/habitica#5490]
Several times recently, I've opened up one of the pulldown menus (Social, the settings gear, etc.), and tried reflexively to dismiss it by clicking in empty space somewhere. That closes modals in Habit and similar menus in many other apps, but here it has no effect. It takes a second of puzzlement before I remember I have to click the menu itself again to close it.

Getting a click-away dismiss would be nice and spare me a few senior moments.
##

**Comment by deilann** [comment]
There's already a ticket for this, and I recall the answer was it's really
useful to mobile and people who have trouble with accurate clicking.

On Thu, Jun 25, 2015 at 10:44 AM, Sabe Jones notifications@github.com
wrote:

> Several times recently, I've opened up one of the pulldown menus (Social,
> the settings gear, etc.), and tried reflexively to dismiss it by clicking
> in empty space somewhere. That closes modals in Habit and similar menus in
> many other apps, but here it has no effect. It takes a second of puzzlement
> before I remember I have to click the menu itself again to close it.
> 
> Getting a click-away dismiss would be nice and spare me a few senior
> moments.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/issues/5490.

## 

Currently Reading
https://www.goodreads.com/user/email_signature_destination/35215676?utm_medium=reading_link&utm_source=email_signature
[image: Book Cover]
https://www.goodreads.com/user/email_signature_destination/35215676?utm_medium=cover&utm_source=email_signature

**Comment by Alys** [comment]
Sort of related https://github.com/HabitRPG/habitrpg/issues/3783 and in one of the tickets linked from that [I wrote](https://github.com/HabitRPG/habitrpg/issues/3978#issuecomment-54349419) "I believe I read somewhere that this is deliberate behaviour intended for usability for users in certain situations. Mobile devices and/or users with poor mouse control maybe??? I can't find that ticket though."

I still can't find it. :-/

**Comment by crookedneighbor** [comment]
@Alys The mobile devices use case doesn't matter for us, since a separate menu system appears for mobile devices with a close button. 

Not sure about poor mouse control. Seems logical, but I've never heard that before.

**Comment by lemoness** [comment]
For what is worth, I have seen this menu behavior cause frustration with
users during my onboarding tests - but because I was under the impression
that it was an accessibility feature, I hadn't filed a ticket about it.

However - and again, I can't find the ticket, so I may be misremembering -
I believe that the arguments for accessibility were that the menu should
not disappear when you stop hovering, as that was difficult for people.
Closing it when you click outside the menu might not be as bad? But I can
still imagine that being potentially frustrating.

In short, if there is an accessible way to improve this UX, that would be
great, but I don't want us to take away an accessible feature.

**Comment by deilann** [comment]
Closing it when you click outside the menu is not that helpful to those
with fine motor skill issues because if you miss a link entirely, it will
go away.

However, closing it when moving to another page is probably less
problematic.

On Fri, Jun 26, 2015 at 6:47 PM, lemoness notifications@github.com wrote:

> For what is worth, I have seen this menu behavior cause frustration with
> users during my onboarding tests - but because I was under the impression
> that it was an accessibility feature, I hadn't filed a ticket about it.
> 
> However - and again, I can't find the ticket, so I may be misremembering -
> I believe that the arguments for accessibility were that the menu should
> not disappear when you stop hovering, as that was difficult for people.
> Closing it when you click outside the menu might not be as bad? But I can
> still imagine that being potentially frustrating.
> 
> In short, if there is an accessible way to improve this UX, that would be
> great, but I don't want us to take away an accessible feature.
> 
> On Fri, Jun 26, 2015 at 4:56 AM, Blade Barringer <notifications@github.com
> 
> > wrote:
> > 
> > @Alys https://github.com/Alys The mobile devices use case doesn't
> > matter for us, since a separate menu system appears for mobile devices
> > with
> > a close button.
> > 
> > Not sure about poor mouse control. Seems logical, but I've never heard
> > that before.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/HabitRPG/habitrpg/issues/5490#issuecomment-115655428
> > .
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/issues/5490#issuecomment-115935315.

## 

Currently Reading
https://www.goodreads.com/user/email_signature_destination/35215676?utm_medium=reading_link&utm_source=email_signature
[image: Book Cover]
https://www.goodreads.com/user/email_signature_destination/35215676?utm_medium=cover&utm_source=email_signature

**Comment by TheHollidayInn** [comment]
So, I have found two ways to easily do this.

The first one is to add a click event to the body/wrapper tag that will dismiss the menu when clicked.

Or, since we seem to be using the ui-bootrstrap directives, we could just change our dropdown menus to use the bootstrap dropdown directive: https://angular-ui.github.io/bootstrap/#/dropdown .

Is there a preference?

**Comment by Alys** [comment]
@TheHollidayInn So that change would allow the menu to close when you clicked outside the menu? As dailann says, that's not necessarily desirable.

@lemoness @AnitaYCheng We should probably work out exactly what behaviour we need for the menus for accessibility and usability.

I recommend that no PRs are submitted until that's done.

**Comment by TheHollidayInn** [comment]
@alys Ah, I misunderstood. I will wait for further discussion then.

**Comment by aipsensmall** [comment]
Clicking outside a menu is indeed normal behavior for most apps. I wouldn't call this an accessibility issue or something needed by users with poor mouse control. This has just become expected behavior in most places. For the record, I have great mouse control and these menus still bother me. :)
Another option is to close menus after a short delay since none of them have very many items.

An even weirder and more annoying behavior of these menus is that clicking on another page while a menu is open does not close it. For example, click on 'Inventory', then open the menu for 'Social', then click on 'Tasks'. There is no reason the 'Social' menu should still be open.

**Comment by Alys** [comment]
@aipsensmall _"I have great mouse control and these menus still bother me"_  The menus bother me a little too if I open one by accident, but the point is that we are attempting to assist people who have **poor** mouse control. Our belief so far is that it's better for accessibility to leave the menus open until an item is clicked or a different menu is opened or the menu is closed manually. I'd prefer that we didn't change the behaviour until we are certain that it's not needed for accessibility.

Closing the menu when moving to another page is less problematic, as has been mentioned above. If someone wanted to make that change now, I think that would be acceptable.

**Comment by sidawson** [comment]
For consistency (and a less confusing UX), and based on the above logic, we should also make the User Popups (clicking another person's icon when in a party - the window that comes up with their stats) **not** disappear when clicking off it.

At present if you click off that window it disappears (typical behaviour for any other site on the net). However, if that's not the way things are done on Habitica (for accessibility reasons), then it should at least be consistent.

**Comment by crookedneighbor** [comment]
I disagree. The surface area in the menu popups is significantly smaller than the surface area of the modals.

**Comment by sidawson** [comment]
Is there an agreed upon pixel area size transition point where we move from one type of behaviour to the other? 

And should that be pixels or ems? (given that - particularly those with accessibility issues - people will tend to adjust their screen+browser and thus pixel resolution to assist themselves - eg View Zoom In/Out in your browser menu)

**Comment by Alys** [comment]
@sidawson I don't think it's necessary to tie it down to particular pixel dimensions. We can make case-by-case decisions.

**Comment by sidawson** [comment]
Another possibility (retaining existing functionality for accessibility requiring users, while making it less confusing for everyone else) would be to have an "accessibility" option (or section) in the site options.

This would free us up to break functionality all over the site in all sorts of interesting ways - making accessibility in general MUCH better for those users requiring it, while generally tightening up behaviour and making things much less confusing for the vast majority of (non accessibility requiring) users.

On a related note, there's also a world of difference between
1. Mouse off the menu to close it
2. Click off the menu to close it

Case 1 is a noted difficult use-case in multi-level (not an issue here) or large menus (also not an issue here), but can be generally fiddly for those without precise mouse control (old mice, fiddly trackpads, plus those with general motor control/coordination issues etc). I'm personally fine with it, but in general I'd agree re avoiding that behaviour.

Case 2 is ONLY going to be a problem for those who regularly mis click, which, I'd strongly suggest is a VERY small percentage of users. 

These "inaccurate clicking users" are also going to:
- be unable to click checklist items
- regularly delete tasks instead of editing them
- click incorrect habits
- mis-assign attributes (with no undo)
  (and so on - ie anywhere on the site that has click zones similarly sized and located)

In other words, the people for whom it's necessary to leave the menu open if they click off it are people for whom the majority of Habitica (as it currently stands) will be utterly unusable, or so frustrating as to be unusable.

**In short: leaving it the way it is, is an inferior experience for everyone that practically benefits no-one.**

Putting it toggled by an accessibility option allows us to keep that functionality if anyone actually prefers/needs it, but also do things like change css to put checklists and other click zones farther apart, make those click zones larger, add yes/no modals in case of incorrect attribute selection and otherwise make Habitica as a whole MUCH simpler for those genuinely struggling to use it due to motor control issues.

**Comment by tobiasdroste** [comment]
I think sidawson is completely right. Look at all the big sites like Facebook, Google, etc. They all use this "click-away"-behavior to close menus. I guess the vast majority of people certainly has the skill to click the links in the menus. If they aren't able to click those menu links they won't be able to use most of the sites on the internet. In such a case they will probably just zoom in on the whole site as they will be having problems at other points on the site too like the checklists.

**Comment by keysmashes** [comment]
Was any decision reached on this? I'd personally like to see click-outside-to-close implemented, and would be happy to investigate it myself (life permitting).

**Comment by SabreCat** [comment]
@anowlcalledjosh Help is welcome on this per the label! However, it's likely to be rendered moot by a frontend redesign we're working on, so if it proves to be a difficult task, it may not be worth sinking a lot of hours into.

**Comment by keysmashes** [comment]
I've converted the dropdowns at the top to ui-bootstrap dropdowns (note to self: [version 0.13.0](https://angular-ui.github.io/bootstrap/versioned-docs/0.13.0/#/dropdown)), and they seem to be working OK. They look almost exactly like the original menus:

![image](https://cloud.githubusercontent.com/assets/9433472/23896813/98566040-08a2-11e7-8a33-14eedb63509a.png)

However, the menus on the right (settings, audio etc.) are causing me some grief. On desktop, they are too far to the right:

![image](https://cloud.githubusercontent.com/assets/9433472/23896477/74a11484-08a1-11e7-8d9a-ad249f195374.png)

Overriding Bootstrap's `.dropdown-menu {left: 0}` with `left: auto` fixes it:

![image](https://cloud.githubusercontent.com/assets/9433472/23896519/a0c8d830-08a1-11e7-9791-699947490939.png)

but this breaks the mobile version (original vs. `left: auto`):

![image](https://cloud.githubusercontent.com/assets/9433472/23896559/c404ee1a-08a1-11e7-9825-45cc1172592e.png)

It seems like it's possible to avoid this by having `left: auto` be overridden itself by the CSS rules for small screens, but I can't work out where to put `left: auto` to make this work. Presumably it goes somewhere in `website/client-old/css/menu.styl`? I can't keep track of which classes are what :(

**Comment by keysmashes** [comment]
Nope, I needed to tweak `$hrpg-modal-dropdown > div` in `global-modules.styl`. I'll have a pull request open soon™.
